fix sgf canonical-data (#1887)
exercism/problem-specifications: exercises/sgf-parsing/description.md

Here `AB` (add black) is used to add three black stones to the board.

Being that parentheses, square brackets and semicolons are integral parts of
the SGF format they must be escaped by preceding the character with a backslash
(i.e. `\(`, `\;` and etc.) when found in text. Some other characters that

> require escaping are newlines, tabs and backslashes (`\n`, `\t`, `\\`). In

Reviewer: These four lines are relevant to someone who wants to output a string in the SGF format. Students are not asked to do that (students are parsing, not outputting), so this information is more information for them to read which ultimately does not help them.

However, it is useful to discuss how to deal with escape sequences when parsing. So perhaps instead just a short bit like "characters preceded by a backslash (the escape character) appear in the output as-is, without the backslash" ?

Finally, if these four lines are instead kept as-is, I cannot verify their truthfulness, because they contradict what's said in https://www.red-bean.com/sgf/sgf4.html

> Escaping: "\" is the escape character. Any char following "\" is inserted verbatim (exception: whitespaces still have to be converted to space!). Following chars have to be escaped, when used in Text: "]", "\" and ":" (only if used in compose data type).

Author: 1. I can fix the language to explain how/when escaped characters will be encountered instead of saying that chars need to be escaped.
2. I can update the content to conform to what the spec says
3. one thing I don't find clear in the spec is how to handle line-breaks, the expectation in this exercise is clearly to replace the C# string `"\\n"` with the char `\n` so I'm going to go with that for now

Author: updated, what do you think?

Here `AB` (add black) is used to add three black stones to the board.

According the the SGF specification `\` is the escape character. When `\` is
encountered during parsing, it indicates that the following character should be
printed as is. The exception to this is whitespace, with newline, carriage
return and tab encoded as `"\n"`, `"\r"` and `"\t"` respectively. When parsing
you can expect some characters to _always_ be escaped, including parentheses,

> square brackets, semicolons and colons.

Reviewer: I'm unsure about the sentence beginning with "When parsing you can expect..." for two reasons:

1. https://www.red-bean.com/sgf/sgf4.html does not say that parentheses or semicolons will always be escaped, so if we make this claim, we are going against the specification... but that is okay! It's okay for this exercise's description to be **more restrictive** than the specification. It just means that we'll never test some inputs that would be legal, and that's fine. If we were **less restrictive** than the specification, then we would test inputs that would be illegal, and we should never do that. So if we're keeping this sentence as-is, let us just agree here that we have decided to deliberately be more restrictive than the specification, to make things easier on the students. Please confirm.
2. It's not clear to me that the student's job is made easier by knowing this. I guess it means that the student can immediately search for the next non-text character they expect to see (that would typically be `]` in the inputs we're testing), and if they find it's been escaped, they can just search for the next one, repeating until they find a non-escaped one? Sure I guess, that's fine. I suppose it doesn't hurt.

Author: 1. yes the specification didn't say that semicolon or parenthesis must be escaped, but I don't understand how you can practically include those chars in the text without escaping them (you may be able to so closing parens without escaping but not semicolon or open parens) so I assumed its a typo. the spec does say that colons must be escaped which I suspect is a type intending semicolon but I could be wrong there so I included colons.
2. it seems to me the student needs to know how to handle these chars when encountered in text. if in theory these chars could appear unescaped it make the task much more complex so I think this information is very helpful

Reviewer: Huh, sounds like they would be interesting test cases to add, then... This statement is **not** a request that the test case should be added as a part of this PR, it was just a statement in passing of what might make a good future addition.

The reason you can have an unescaped semicolon inside a Text is that Texts are only closed by the specific character `]`, so if you've encountered a `;` before a `]` you know the semicolon is part of the Text and not the semicolon that indicates a new node. Same reasoning for the other characters.

Author: i see your point

When parsing SGF files whitespace gets special handling, with the exception of
newlines any whitespace character encountered is replaced with a single space
character irrespective of if the original character is found in its escaped

> from or its unescaped form.

Reviewer: ```suggestion
form or its unescaped form.
```

> from or its unescaped form.

Reviewer: While I think the idea of this distinction between "its escaped form or its unescaped form" contradicts https://www.red-bean.com/sgf/sgf4.html, fixing this would be a larger undertaking and also involves changing the case currently in the canonical-data.json, so I'll be silent about it for now and refrain from making that a part of this PR.

Author: I thought it was consistent with the spec.

The spec is not explicit but it states
- "White spaces other than linebreaks are converted to space"
- "Any char following "\" is inserted verbatim (exception: whitespaces still have to be converted to space!)."
- "Following chars have to be escaped, when used in Text: "]", "\" and ":" (only if used in compose data type)."

which seems to indicate:
1. that escaped WS is interpreted as WS (so`"\t"` is a tab and not an escaped `t`) and should be replaced with a space character.
2. that whitespace does not have to be escaped so unescaped whitespace would also need to be converted to spaces (save for line break)

Reviewer: You have already pointed out the crucial statement: "Any char following `"\"` is inserted verbatim". Therefore, the SGF parser, when seeing the sequence of characters `\t`, should not emit a tab or a space but just a `t`.

Since we have a test in the canonical-data.json that does otherwise, **our canonical-data.json is wrong**. But, we can fix that later - since the test has been wrong for years, no additional harm is caused by it remaining wrong for some time longer.

The improvement to the canonical-data.json would probably be having individual tests for some of the individual required behaviours (replace whitespace with a space, replace an escaped newline with nothing at all, all other escaped characters).

So, let's not worry about it for this PR and just stick with what's there.

Author: `"Any char following "\" is inserted verbatim (exception: whitespaces still have to be converted to space!)."` you need to read what they have in the parentheses, if `"\t"` was to be replaced with `"t"` then how do you understand that "exception"?

Reviewer: A backslash followed by a tab should become a space. And a backslash followed by a t should become a t.

Reviewer: Important to distinguish between:

* strings as they appear in a in JSON or a given programming language's string literal
* strings as they are presented to the SGF parser

If the string presented to the SGF parser contains a backslash followed by a tab (which will typically be written in JSON and/or in a language's string literal as `"\\\t"` that should be converted to a space, because "(exception: whitespaces still have to be converted to space)".

If the string presented to the SGF parser contains a backslash followed by a t (which will typically be written in JSON and/or in a language's string literal as "\\t" that should be converted to a t, because "Any char following `"\"` is inserted verbatim"

Reviewer: Important to distinguish between:

* strings as they appear in a in JSON or a given programming language's string literal
* strings as they are presented to the SGF parser

If the string presented to the SGF parser contains a backslash followed by a tab (which will typically be written in JSON and/or in a language's string literal as `"\\\t"` that should be converted to a space, because "(exception: whitespaces still have to be converted to space)".

If the string presented to the SGF parser contains a backslash followed by a t (which will typically be written in JSON and/or in a language's string literal as `"\\t"` that should be converted to a t, because "Any char following `"\"` is inserted verbatim"

Reviewer: If the string presented to the SGF parser contains a tab (written as "\t"), that should be converted to a space because "White spaces other than linebreaks are converted to space"

Here `AB` (add black) is used to add three black stones to the board.

According the the SGF specification `\` is the escape character. When `\` is
encountered during parsing it indicates that the following character should be

> printed as is. The exception to this is whitespace, as is common with many
> encoding schemas the character sequences `"\n"`, `"\r"` and `"\t"` are the
> escaped forms of the newline, carriage return and space characters
> respectively. When parsing you can expect some characters to always be escaped
> including parentheses, square brackets, semicolons and colons.

Reviewer 2: Maybe shorten this a bit.

```suggestion
printed as is. The exception to this is whitespace, with newline, carriage return and 
tabs encoded as `"\n"`, `"\r"` and `"\t"` respectively. When parsing you can expect some characters to _always_ be escaped, including parentheses, square brackets, semicolons 
and colons.
```

There was also a small mistake, where is said the `\t` encoded spaces.
